### PR TITLE
`read`, `contains`, `create`, `delete` for local docs

### PR DESCRIPTION
**Pull request by tomblench**
## Checklist

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)
- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`|`CHANGELOG.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description

See #589 - allow users to interact with local docs via `read`, `contains`, `create`, and `delete`.

Note that other public API methods explicitly _do not_ interact with local docs (eg `changes`).

## Approach

Users can interact with local documents by using existing methods but prefixing the document ID with `_local/`. Where a document is returned, it retains its `_local/` prefix.

## Schema & API Changes

"No change"

## Security and Privacy

"No change"

## Testing

Added `LocalDocumentCrud` test class. This is fairly comprehensive and should also help document the expected behaviour for some corner cases.

## Monitoring and Logging

Eventbus events are fired for local documents in a manner analogous to non-local documents (there are assertions for these in the tests).

### CODE REVIEW

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java [hunk 1]
```diff
@@ -1016,6 +1053,14 @@ public DocumentRevision delete(final DocumentRevision rev) throws
     public List<DocumentRevision> delete(final String id)
             throws DocumentStoreException {
         Misc.checkNotNull(id, "ID");
+        // check for local prefix:
+        // there are two reasons not to delete local documents here:
+        // 1) "delete all leaf nodes" has no meaning for local documents
+        // 2) DeleteAllRevisionsCallable can throw DocumentNotFoundException which would mean a
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java]: I'm not sure I quite follow this.
`db.delete("_local/mydoc")` seems a lot more intuitive than `db.delete(new DocumentRevisions("_local/mydoc",null))` so it might be nice to have this even though the return of an empty list is a little weird for a local doc.
Are you saying that we don't expose `DocumentNotFoundException` on the signature of this method, although we do for `delete(DocumentRev)`?
It seems to me that `DeleteAllRevisionsCallable` throwing a `DocumentNotFoundException` results in a `DocumentStoreException`, but you're reluctant to have this same behaviour for deleting the local doc?

Comment (tomblench) [reply cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java]: OK, I've changed this in 5fac0be so that this flavour also deletes local documents, and changed the method signature as discussed

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java [hunk 2]
```diff
@@ -942,6 +965,10 @@ public DocumentRevision update(final DocumentRevision rev)
         Misc.checkArgument(rev.isFullRevision(), "Projected revisions cannot be used to " +
                 "create documents");
 
+        if (rev.getId().startsWith(CouchConstants._local_prefix)) {
+            throw new IllegalArgumentException("Use create(final DocumentRevision rev) create or update local documents");
```
[the remark this answers is not in the document]
Comment (tomblench) [reply cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java]: Done in 227e1c3

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/LocalDocument.java [hunk 3]
```diff
@@ -15,8 +15,15 @@
 package com.cloudant.sync.documentstore;
 
 /**
- * A local Document. {@code LocalDocument}s do not have a history, or the concept of revisions
+ * <p>
+ *     <b>Note:</b> this class is deprecated and will be moved to an internal package in a future
+ *     release.
```
[the remark this answers is not in the document]
Comment (tomblench) [reply cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/LocalDocument.java]: Done in 6df3a64

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/Database.java [hunk 4]
```diff
@@ -296,7 +310,9 @@ DocumentRevision update(DocumentRevision rev) throws ConflictException,
             AttachmentException, DocumentStoreException, DocumentNotFoundException;
 
     /**
-     * <p>Deletes a document from the DocumentStore.</p>
+     * <p>Deletes a document from the DocumentStore; or delete a
+     * <a href="https://couchdb.readthedocs.io/en/stable/api/local.html">local document</a>,
+     * if {@code documentId} is prefixed with {@code _local/}.</p>
      *
      * <p>This operation leaves a "tombstone" for the deleted document, so that
```
[the remark this answers is not in the document]
Comment (tomblench) [reply cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/Database.java]: Done in 0695339

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/Database.java [hunk 5]
```diff
@@ -284,6 +296,8 @@ DocumentRevision create(DocumentRevision rev) throws AttachmentException,
      * {@link com.cloudant.sync.event.notifications.DocumentUpdated DocumentUpdated}
      * event is posted on the event bus.</p>
      *
+     * <p><b>Note:</b> to update local documents, call {@link #create(DocumentRevision)}</p>
```
[the remark this answers is not in the document]
Comment (tomblench) [reply cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/Database.java]: The documentation in 227e1c3 now infers this, do you think it's clear enough?
